compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,265 @@
+//====---------------------------------------------------------------------===//
+// TODO: Move this to a common place.
+//====---------------------------------------------------------------------===//
+
+static constexpr int kNumBitsPerCopy = 32;
+
+static LogicalResult
+distributeLinalgCopyToThreads(RewriterBase &rewriter, linalg::CopyOp copy,
+                              ArrayRef<int64_t> workgroupSize,
+                              ArrayRef<int64_t> subgroupSize) {
+  LDBG("==== distributing op:\n"
+       << "\t" << *copy);
+
+  Location loc = copy.getLoc();
+  MLIRContext *context = rewriter.getContext();
+
+  auto getTotalSize = [](ArrayRef<int64_t> sizes) {
+    return std::accumulate(sizes.begin(), sizes.end(), 1,
+                           std::multiplies<int64_t>());
+  };
+
+  // The linalg.copy we are dealing with represents a region we need to copy to
+  // workgroup memory. Assume there are N threads in the workgroup, then there
+  // are `num_subgroups = N / gpu.subgroup_size` subgroups in the workgroup.
+  //
+  // So we are slicing up the target memref into `num_subgroups` consecutive
+  // slices, and threads in the same subgroup will copy their slice to workgroup
+  // memory slice.
+
+  // Get the copy size:
+  auto copyMemRefType = cast<MemRefType>(copy.getOperand(1).getType());
+  if (!memref::isStaticShapeAndContiguousRowMajor(copyMemRefType)) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: copy to non-static or non-contiguous, "
+              "non-row major memref.");
+  }
+  auto rank = copyMemRefType.getRank();
+  SmallVector<OpFoldResult> tileSize(rank - 1, rewriter.getIndexAttr(1));
+
+  int64_t elementBitWidth = copyMemRefType.getElementTypeBitWidth();
+  if (kNumBitsPerCopy % elementBitWidth != 0) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: preferred copy size is not a multiple of "
+              "element bit width.");
+  }
+  int64_t elementsPerCopy = kNumBitsPerCopy / elementBitWidth;
+
+  // divide the copy by subgroup:
+  assert(subgroupSize.size() == 1); // only 1-D
+  assert(workgroupSize[0] % subgroupSize[0] == 0);
+
+  int64_t numSubgroups = workgroupSize[0] / subgroupSize[0];
+  int64_t totalCopySize = getTotalSize(copyMemRefType.getShape());
+  int64_t totalCopySizePerSubgroup = totalCopySize / numSubgroups;
+  int64_t numCopiesPerThread =
+      (totalCopySizePerSubgroup / elementsPerCopy) / subgroupSize[0];
+  int64_t residualElements =
+      totalCopySizePerSubgroup % (subgroupSize[0] * elementsPerCopy);
+
+  LDBG("-- elementsPerCopy: " << elementsPerCopy << "\n");
+  LDBG("-- workgroupSize: " << workgroupSize[0] << "\n");
+  LDBG("-- numSubgroups: " << numSubgroups << "\n");
+  LDBG("-- totalCopySize: " << totalCopySize << "\n");
+  LDBG("-- totalCopySizePerSubgroup: " << totalCopySizePerSubgroup << "\n");
+  LDBG("-- numCopiesPerThread: " << numCopiesPerThread << "\n");
+  LDBG("-- residualElements: " << residualElements << "\n");
```

Comment:
nit: delete the `"\n"` from log because this is why we introduce LDBG. Same comment for other logging.